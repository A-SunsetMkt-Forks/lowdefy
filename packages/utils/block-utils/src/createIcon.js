/*
  Copyright 2020-2022 Lowdefy, Inc

  Licensed under the Apache License, Version 2.0 (the "License");
  you may not use this file except in compliance with the License.
  You may obtain a copy of the License at

      http://www.apache.org/licenses/LICENSE-2.0

  Unless required by applicable law or agreed to in writing, software
  distributed under the License is distributed on an "AS IS" BASIS,
  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  See the License for the specific language governing permissions and
  limitations under the License.
*/

import React from 'react';
import classNames from 'classnames';
import { keyframes } from '@emotion/react';
import { css } from '@emotion/css';
import { omit, type } from '@lowdefy/helpers';
import Icon from '@ant-design/icons';

import blockDefaultProps from './blockDefaultProps.js';
import ErrorBoundary from './ErrorBoundary.js';
import makeCssClass from './makeCssClass.js';

const lowdefyProps = [
  'actionLog',
  'basePath',
  'components',
  'content',
  'eventLog',
  'list',
  'loading',
  'menus',
  'pageId',
  'registerEvent',
  'registerMethod',
  'schemaErrors',
  'user',
  'validation',
];

const spin = keyframes`{
  0% {
    transform: rotate(0deg);
  }
  100% {
    transform: rotate(359deg);
  }
}`;

const spinClass = css`
  animation: ${spin} 2s infinite linear;
`;

const createIcon = (Icons) => {
  const AiOutlineLoading3Quarters = Icons['AiOutlineLoading3Quarters'];
  const AiOutlineExclamationCircle = Icons['AiOutlineExclamationCircle'];

  const IconBlock = ({ blockId, events, methods, properties, ...props }) => {
    const propertiesObj = type.isString(properties) ? { name: properties } : properties;
    const iconProps = {
      id: blockId,
      className: classNames({
        [makeCssClass(propertiesObj.style)]: true,
        [spinClass]: propertiesObj.spin,
      }),
      rotate: propertiesObj.rotate,
      color: propertiesObj.color,
      title: propertiesObj.name,
      size: propertiesObj.size,
      // twoToneColor: propertiesObj.color, // TODO: track https://github.com/react-icons/react-icons/issues/508
      ...omit(props, lowdefyProps),
    };
    let IconComp = Icons[propertiesObj.name];
    if (!IconComp) {
      IconComp = AiOutlineExclamationCircle;
    }
    return (
      <>
        {events.onClick && events.onClick.loading && !propertiesObj.disableLoadingIcon ? (
          <AiOutlineLoading3Quarters {...{ ...iconProps, spin: true }} />
        ) : (
          <ErrorBoundary
            fallback={() => <AiOutlineExclamationCircle {...{ ...iconProps, color: '#F00' }} />}
          >
            <IconComp
              id={blockId}
              onClick={
                events.onClick &&
                (() =>
                  methods.triggerEvent({
                    name: 'onClick',
                  }))
              }
              size={propertiesObj.size}
              title={propertiesObj.title}
              {...iconProps} // spread props for to populate props from parent
            />
          </ErrorBoundary>
        )}
      </>
    );
  };
  const AntIcon = (all) => <Icon component={() => <IconBlock {...all} />} />;
  AntIcon.defaultProps = blockDefaultProps;
  return AntIcon;
};

export default createIcon;
