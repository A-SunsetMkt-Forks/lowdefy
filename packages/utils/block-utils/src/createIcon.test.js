/*
  Copyright 2020-2022 Lowdefy, Inc

  Licensed under the Apache License, Version 2.0 (the "License");
  you may not use this file except in compliance with the License.
  You may obtain a copy of the License at

      http://www.apache.org/licenses/LICENSE-2.0

  Unless required by applicable law or agreed to in writing, software
  distributed under the License is distributed on an "AS IS" BASIS,
  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  See the License for the specific language governing permissions and
  limitations under the License.
*/

import React from 'react';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import createIcon from './createIcon.js';
import makeCssClass from './makeCssClass.js';

const methods = {
  makeCssClass,
  triggerEvent: jest.fn(),
};

beforeAll(() => {
  console.error = () => null;
});

const Icons = {
  AiIcon: (props) => <svg {...props} data-testid="AiIcon"></svg>,
  AiOutlineExclamationCircle: (props) => (
    <svg {...props} data-testid="AiOutlineExclamationCircle"></svg>
  ),
  AiOutlineLoading3Quarters: (props) => (
    <svg {...props} data-testid="AiOutlineLoading3Quarters"></svg>
  ),
  ErrorIcon: () => {
    throw new Error('ErrorIcon');
  },
};

test('Icon default', () => {
  const IconComponent = createIcon(Icons);
  const { container } = render(<IconComponent methods={methods} />);
  expect(container.firstChild).toMatchSnapshot();
});

test('Icon default and id', () => {
  const IconComponent = createIcon(Icons);
  const { container } = render(<IconComponent id="test-id" methods={methods} />);
  expect(container.firstChild).toMatchSnapshot();
});

test('Icon properties.name', () => {
  const IconComponent = createIcon(Icons);
  const { container } = render(
    <IconComponent id="test-id" methods={methods} properties={{ name: 'AiIcon' }} />
  );
  expect(container.firstChild).toMatchSnapshot();
});

test('Icon properties string', () => {
  const IconComponent = createIcon(Icons);
  const { container } = render(
    <IconComponent id="test-id" methods={methods} properties="AiIcon" />
  );
  expect(container.firstChild).toMatchSnapshot();
});

test('Icon properties.spin', () => {
  const IconComponent = createIcon(Icons);
  const { container } = render(
    <IconComponent id="test-id" methods={methods} properties={{ name: 'AiIcon', spin: true }} />
  );
  expect(container.firstChild).toMatchSnapshot();
});

test('Icon properties.style', () => {
  const IconComponent = createIcon(Icons);
  const { container } = render(
    <IconComponent
      id="test-id"
      methods={methods}
      properties={{ name: 'AiIcon', style: { background: 'yellow' } }}
    />
  );
  expect(container.firstChild).toMatchSnapshot();
});

test('Icon properties.name error', () => {
  const IconComponent = createIcon(Icons);
  const { container } = render(
    <IconComponent id="test-id" methods={methods} properties={{ name: 'ErrorIcon' }} />
  );
  expect(container.firstChild).toMatchSnapshot();
});

test('Icon onClick.loading false', () => {
  const IconComponent = createIcon(Icons);
  const { container } = render(
    <IconComponent
      id="test-id"
      methods={methods}
      properties={{ name: 'AiIcon' }}
      events={{ onClick: { loading: false } }}
    />
  );
  expect(container.firstChild).toMatchSnapshot();
  userEvent.click(screen.getByTestId('AiIcon'));
  expect(methods.triggerEvent).toHaveBeenCalledWith({ name: 'onClick' });
});

test('Icon onClick.loading true', () => {
  const IconComponent = createIcon(Icons);
  const { container } = render(
    <IconComponent
      id="test-id"
      methods={methods}
      properties={{ name: 'AiIcon' }}
      events={{ onClick: { loading: true } }}
    />
  );
  expect(container.firstChild).toMatchSnapshot();
});
