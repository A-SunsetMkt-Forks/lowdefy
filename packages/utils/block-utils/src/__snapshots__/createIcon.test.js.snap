// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Icon default 1`] = `
<span
  class="anticon"
  role="img"
>
  <svg
    class="emotion-0"
    data-testid="AiOutlineExclamationCircle"
    id="undefined_id"
  />
</span>
`;

exports[`Icon default and id 1`] = `
<span
  class="anticon"
  role="img"
>
  <svg
    class="emotion-0"
    data-testid="AiOutlineExclamationCircle"
    id="test-id"
  />
</span>
`;

exports[`Icon onClick.loading false 1`] = `
<span
  class="anticon"
  role="img"
>
  <svg
    class="emotion-0"
    data-testid="AiIcon"
    id="test-id"
    title="AiIcon"
  />
</span>
`;

exports[`Icon onClick.loading true 1`] = `
<span
  class="anticon"
  role="img"
>
  <svg
    class="emotion-0"
    data-testid="AiOutlineLoading3Quarters"
    id="test-id"
    title="AiIcon"
  />
</span>
`;

exports[`Icon properties string 1`] = `
<span
  class="anticon"
  role="img"
>
  <svg
    class="emotion-0"
    data-testid="AiIcon"
    id="test-id"
    title="AiIcon"
  />
</span>
`;

exports[`Icon properties.name 1`] = `
<span
  class="anticon"
  role="img"
>
  <svg
    class="emotion-0"
    data-testid="AiIcon"
    id="test-id"
    title="AiIcon"
  />
</span>
`;

exports[`Icon properties.name error 1`] = `
<span
  class="anticon"
  role="img"
>
  <svg
    class="emotion-0"
    color="#F00"
    data-testid="AiOutlineExclamationCircle"
    id="test-id"
    title="ErrorIcon"
  />
</span>
`;

exports[`Icon properties.spin 1`] = `
@keyframes animation-0 {
  {
    -webkit-transform: rotate(0deg);
    -moz-transform: rotate(0deg);
    -ms-transform: rotate(0deg);
    transform: rotate(0deg);
  }

   {
    -webkit-transform: rotate(359deg);
    -moz-transform: rotate(359deg);
    -ms-transform: rotate(359deg);
    transform: rotate(359deg);
  }
}

.emotion-1 {
  -webkit-animation: animation-0 2s infinite linear;
  animation: animation-0 2s infinite linear;
}

<span
  class="anticon"
  role="img"
>
  <svg
    class="emotion-0 emotion-1"
    data-testid="AiIcon"
    id="test-id"
    title="AiIcon"
  />
</span>
`;

exports[`Icon properties.style 1`] = `
.emotion-0 {
  background: yellow;
}

<span
  class="anticon"
  role="img"
>
  <svg
    class="emotion-0"
    data-testid="AiIcon"
    id="test-id"
    title="AiIcon"
  />
</span>
`;
